src/utils/error.ts

```diff
@@ -73,30 +73,67 @@ export class FirebaseError extends Error {
 }
 
 /**
- * Firebase App error code structure. This extends FirebaseError.
+ * A FirebaseError with a prefix in front of the error code.
  *
+ * @param {string} codePrefix The prefix to apply to the error code.
  * @param {string} code The error code.
  * @param {string} message The error message.
  * @constructor
  */
-export class FirebaseAppError extends FirebaseError {
-  constructor(code: string, message: string) {
+class PrefixedFirebaseError extends FirebaseError {
+  constructor(private codePrefix: string, code: string, message: string) {
     super({
-      code: 'app/' + code,
+      code: `${codePrefix}/${code}`,
       message,
     });
+
+    /* tslint:disable:max-line-length */
+    // Set the prototype explicitly. See the following link for more details:
+    // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    /* tslint:enable:max-line-length */
+    (this as any).__proto__ = PrefixedFirebaseError.prototype;
```

Comment:
Should we use `Object.setPrototypeOf()` here as shown in the doc you've linked?

Reply:
This was just copied from elsewhere in the same file: https://github.com/firebase/firebase-admin-node/blob/master/src/utils/error.ts#L53

I applied it to every subclass, however, because the current implementation is buggy in that it actually breaks the inheritance.

I can't really speak to the difference between `(this as any).__proto__` and `Object.setPrototypeOf()` but happy to make the change if someone who knows better chimes in :)